sys/net/crosslayer/ng_pktbuf/ng_pktbuf.c

```diff
@@ -147,60 +168,70 @@ static ng_pktsnip_t *_pktbuf_alloc(size_t size)
     return pkt;
 }
 
-static ng_pktsnip_t *_pktbuf_add_unsafe(ng_pktsnip_t *next, void *data,
+static ng_pktsnip_t *_pktbuf_add_unsafe(ng_pktsnip_t *payload, void *data,
                                         size_t size, ng_nettype_t type)
 {
-    ng_pktsnip_t *snip;
+    ng_pktsnip_t *hdr;
 
     if (size == 0) {
         data = 0;
     }
 
-    snip = (ng_pktsnip_t *)_pktbuf_internal_alloc(sizeof(ng_pktsnip_t));
+    hdr = (ng_pktsnip_t *)_pktbuf_internal_alloc(sizeof(ng_pktsnip_t));
 
-    if (snip == NULL) {
+    if (hdr == NULL) {
         return NULL;
     }
 
-    if (next == NULL || next->data != data) {
-        if (size != 0) {
-            snip->data = _pktbuf_internal_alloc(size);
+    if (payload == NULL || payload->data != data) {
+        if ((size != 0) && (!_pktbuf_internal_contains(data))) {
+            hdr->data = _pktbuf_internal_alloc(size);
 
-            if (snip->data == NULL) {
-                _pktbuf_internal_free(snip);
+            if (hdr->data == NULL) {
+                _pktbuf_internal_free(hdr);
 
                 return NULL;
             }
 
             if (data != NULL) {
-                memcpy(snip->data, data, size);
+                memcpy(hdr->data, data, size);
             }
         }
         else {
-            snip->data = data;
+            if (_pktbuf_internal_contains(data)) {
+                if (!_pktbuf_internal_add_pkt(hdr->data)) {
+                    _pktbuf_internal_free(hdr);
+
+                    return NULL;
+                }
+            }
+
+            hdr->data = data;
         }
+
+        hdr->next = NULL;
+        LL_PREPEND(payload, hdr);
     }
     else {
-        snip->data = data;
+        hdr->next = NULL;
+        hdr->data = data;
 
-        next->size -= size;
-        next->data = (void *)(((uint8_t *)next->data) + size);
+        payload->next = hdr;
```

Comment:
That's why https://github.com/RIOT-OS/RIOT/pull/2563/files#r26119142. This line would be `next->next = pkt`.